### PR TITLE
Create Conda Package for numba-extensions-bf16, and update build system accordingly.

### PR DESCRIPTION
**Pull request by quasiben**
Conda package for bfloat16 (and subsequently for numbast and ast_canopy). This PR includes several temporary changes to the build system:
- libastcanopy.so (previously mistakenly named liblibastcanopy.so) now installed into `SKBUILD_PLATLIB`, not `conda/lib`. `ast_canopy` build system is currently hand crafted so that this shared object is moved into the right location. See `ast_canopy/build.sh` for detail. When doing conda build, the build command is in `meta.yaml::build/script` key. This reuses the `SP_DIR` path provided in the conda build environment, instead of the dynamically determined `stie.getsitepackages()` function in build.sh, which is more robust.

The artifacts of this PR:
- Establishes 2 new CI jobs that builds ast_canopy, numbast and bf16 conda package. Then creates a new conda environment in these jobs, add the `conda_bld` folder as the new channel, and install bf16 conda package. Run the tests in these environments. The 2 CI job are V100 and A100 machines respectively. Hopefully they can replace our current test system. Provided we setup a S3 bucket to upload the conda build artifacts.
- The above verifies certain capability of our build system supporting conda package delivery, but with some drawbacks due to manual installation of `libastcanopy.so`. I'm working to renovate the ast_canopy's build system so that `libastcanopy.so` installation is managed by cmake, and `pylibastcanopy` discovers it automatically. This makes sure that scikit-build-core and conda build all recognizes this dependency.

TODO work:

- [ ] Refactor build system to reduce the manual wrangling needed in this PR. Shown in #52 
- [ ] Deprecate the python test CI job and use the conda build job instead.

### CODE REVIEW

File: .github/actions/compute-matrix/action.yaml
Comment (isVoid) [inline .github/actions/compute-matrix/action.yaml:1]: We can remove this file because the build instances are manually setup by hand.

File: .github/workflows/build.yaml [hunk 1]
```diff
@@ -0,0 +1,26 @@
+
```
Comment (isVoid) [inline .github/workflows/build.yaml:1]:
```suggestion
```

File: .github/workflows/build.yaml [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+# SPDX-FileCopyrightText: Copyright (c) 2024 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+name: Build CONDA PKG
+
+on:
```
Comment (isVoid) [inline .github/workflows/build.yaml:7]: Do we want to setup a nightly build?

File: .github/workflows/conda-python-build.yaml [hunk 3]
```diff
@@ -42,56 +48,18 @@ permissions:
...
       strategy:
         fail-fast: false
-        matrix: ${{ fromJSON(needs.compute-matrix.outputs.MATRIX) }}
-      runs-on: "linux-${{ matrix.ARCH }}-${{ inputs.node_type }}"
+      runs-on: linux-${{ inputs.CPU }}-gpu-${{ inputs.GPU }}-latest-1
       env:
         RAPIDS_ARTIFACTS_DIR: ${{ github.workspace }}/artifacts
       container:
-        image: rapidsai/ci-conda:cuda${{ matrix.CUDA_VER }}-${{ matrix.LINUX_VER }}-py${{ matrix.PY_VER }}
+        image: rapidsai/ci-conda:cuda12.2.2-ubuntu22.04-py3.10
```
Comment (isVoid) [inline .github/workflows/conda-python-build.yaml:58]: The conda test below more or less depends on the fact that this image uses python 3.10, and that it only tests python 3.10 builds. Which is fine for now, given that our current python tests also only tests 3.10.

File: .pre-commit-config.yaml [hunk 4]
```diff
@@ -3,7 +3,6 @@ repos:
 -   repo: https://github.com/pre-commit/pre-commit-hooks
     rev: v2.3.0
     hooks:
-    -   id: check-yaml
```
Comment (isVoid) [inline .pre-commit-config.yaml:6]: Why was this removed?

File: LICENSE [hunk 5]
```diff
@@ -178,15 +178,15 @@
...
+      boilerplate notice, with the fields enclosed by brackets "{}"
       replaced with your own identifying information. (Don't include
       the brackets!)  The text should be enclosed in the appropriate
       comment syntax for the file format. We also recommend that a
       file or class name and description of purpose be included on the
       same "printed page" as the copyright notice for easier
       identification within third-party archives.
 
-   Copyright 2024 Michael Yh Wang
+   Copyright 2018 NVIDIA Corporation
```
Comment (isVoid) [inline LICENSE:189]:
```suggestion
   Copyright 2024 NVIDIA Corporation
```

File: ast_canopy/VERSION
Comment (isVoid) [inline ast_canopy/VERSION:1]: Was this file used anywhere at all?

File: ci/run_tests.py [hunk 6]
```diff
@@ -0,0 +1,75 @@
+import click
+import subprocess
+import os
+
+import pytest
+
+
+def run_pytest(lib, test_dir):
```
Comment (isVoid) [inline ci/run_tests.py:8]: Add documentation

File: ci/run_tests.py [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+
+
+@click.command()
+@click.option("--ast_canopy", is_flag=True)
+@click.option("--numbast", is_flag=True)
+@click.option("--bf16", is_flag=True)
+@click.option("--fp16", is_flag=True)
+@click.option("--curand_device", is_flag=True)
+@click.option("--all_tests", is_flag=True)
+def run(
```
Comment (isVoid) [inline ci/run_tests.py:36]: Add documentation and help text for each of these options and the command line interface.

File: ci/test_conda.sh [hunk 8]
```diff
@@ -0,0 +1,48 @@
+#!/bin/bash
+# Copyright (c) 2024, NVIDIA CORPORATION
```
Comment (isVoid) [inline ci/test_conda.sh:2]: Use standard spdx license header.

File: conda/recipes/ast_canopy/conda_build_config.yaml [hunk 9]
```diff
@@ -0,0 +1,22 @@
+c_compiler_version:
```
Comment (isVoid) [inline conda/recipes/ast_canopy/conda_build_config.yaml:1]: missing license

File: conda/recipes/ast_canopy/meta.yaml [hunk 10]
```diff
@@ -0,0 +1,56 @@
+# Copyright (c) 2018-2024, NVIDIA CORPORATION.
```
Comment (isVoid) [inline conda/recipes/ast_canopy/meta.yaml:1]: spdx license

File: conda/recipes/bf16/conda_build_config.yaml [hunk 11]
```diff
@@ -0,0 +1,22 @@
+c_compiler_version:
```
Comment (isVoid) [inline conda/recipes/bf16/conda_build_config.yaml:1]: missing license

File: conda/recipes/bf16/meta.yaml [hunk 12]
```diff
@@ -0,0 +1,59 @@
+# Copyright (c) 2018-2024, NVIDIA CORPORATION.
```
Comment (isVoid) [inline conda/recipes/bf16/meta.yaml:1]: spdx license

File: conda/recipes/numbast/conda_build_config.yaml [hunk 13]
```diff
@@ -0,0 +1,22 @@
+c_compiler_version:
```
Comment (isVoid) [inline conda/recipes/numbast/conda_build_config.yaml:1]: missing license

File: conda/recipes/numbast/meta.yaml [hunk 14]
```diff
@@ -0,0 +1,51 @@
+# Copyright (c) 2018-2024, NVIDIA CORPORATION.
```
Comment (isVoid) [inline conda/recipes/numbast/meta.yaml:1]: spdx license

File: numba_extensions/bf16/LICENSE
Comment (isVoid) [inline numba_extensions/bf16/LICENSE:1]: Can remove because we are sharing repo_root/LICENSE now.

**Review by isVoid — CHANGES_REQUESTED** [review]

File: .pre-commit-config.yaml (hunk 4 above)
Comment (isVoid) [reply .pre-commit-config.yaml:6]: Because we are using yaml templates in meta.yaml, check-yaml isn't powerful enough to deal with these files. We should look into alternatives.

File: ast_canopy/VERSION
Comment (isVoid) [reply ast_canopy/VERSION:1]: Yes, this is loaded in `conda/recipe/ast_canopy/meta.yaml`.

**Review by isVoid — APPROVED** [review]
Approving pending two questions above.

File: ci/run_tests.py
Comment (quasiben) [inline ci/run_tests.py:1]: Good idea to refactor the testing into a single file

File: .github/workflows/build.yaml (hunk 2 above)
Comment (quasiben) [reply .github/workflows/build.yaml:6]: The cadence of development here will probably be slower than pynjitlink, for example.  So I don't think we need nightly builds (at least not yet).  Instead, I think we should manually do release for a couple reasons:
1. Let's effort now and quicker to ship an artifact to NeMo folks
2. We should learn a bit more about the NVIDIA channel before uploading nightly

Again, we can draw for pynjitlink for managing releases -- I think it relies on GH actions and manual tagging for uploading packages:
- https://github.com/rapidsai/pynvjitlink/pull/74
